src/main/java/bio/terra/service/snapshot/SnapshotService.java

```diff
@@ -633,12 +649,157 @@ public Snapshot makeSnapshotFromSnapshotRequest(SnapshotRequestModel snapshotReq
         .tags(TagUtils.sanitizeTags(snapshotRequestModel.getTags()));
   }
 
-  private void createSnapshotTablesFromDatasetAsset(
-      String datasetName, String assetName, SnapshotSource snapshotSource, Snapshot snapshot) {
-    Dataset queryDataset = datasetService.retrieveByName(datasetName);
-    AssetSpecification queryAssetSpecification = getAssetByNameFromDataset(queryDataset, assetName);
-    snapshotSource.assetSpecification(queryAssetSpecification);
-    conjureSnapshotTablesFromAsset(snapshot, snapshotSource);
+  private SnapshotAccessRequestResponse getById(UUID accessRequestId) {
+    return snapshotRequestDao.getById(accessRequestId);
+  }
+
+  @VisibleForTesting
+  AssetSpecification buildAssetFromSnapshotAccessRequest(
+      Dataset dataset, SnapshotAccessRequestResponse snapshotRequestModel) {
+    // build asset model from snapshot request
+    AssetModel assetModel = new AssetModel();
+    assetModel.name("snapshot-by-request-asset");
+    assetModel.rootTable("person");
+    assetModel.rootColumn("person_id");
+    // Manually add dictionary tables, leave columns empty to return all columns
+    assetModel.addTablesItem(new AssetTableModel().name("person"));
+    assetModel.addTablesItem(new AssetTableModel().name("concept"));
+
+    // Build asset tables and columns based on the concept sets included in the snapshot request
+    List<SnapshotBuilderTable> tables = pullTables(snapshotRequestModel);
+    tables.forEach(
+        table -> {
+          assetModel.addTablesItem(
+              new AssetTableModel().name(table.getDatasetTableName()).columns(table.getColumns()));
+          assetModel.follow(table.getRelationships());
+        });
+
+    // Make sure we just built a valid asset model
+    dataset.validateDatasetAssetSpecification(assetModel);
+
+    // convert the asset model to an asset specification
+    return dataset.getNewAssetSpec(assetModel);
+  }
+
+  @VisibleForTesting
+  List<SnapshotBuilderTable> pullTables(SnapshotAccessRequestResponse snapshotRequestModel) {
+    var valueSets = snapshotRequestModel.getSnapshotSpecification().getValueSets();
+    var valueSetNames = valueSets.stream().map(SnapshotBuilderFeatureValueGroup::getName).toList();
+
+    Map<String, SnapshotBuilderTable> tableMap = populateManualTableMap();
+
+    return valueSetNames.stream()
+        .map(
+            name -> {
+              if (tableMap.containsKey(name)) {
+                return tableMap.get(name);
+              } else {
+                throw new IllegalArgumentException("Unknown value set name: " + name);
+              }
+            })
+        .toList();
+  }
+
+  private Map<String, SnapshotBuilderTable> populateManualTableMap() {
```

Comment:
I wonder if this should be in a separate class so it's clear that it's axin specific?

I guess it won't be around too long so maybe it's OK to leave it here.

Reply:
Is this AXIN specific or just OMOP specific?

Reply:
So just to clarify again, all of the information from this map will be stored in the settings for each snapshot so this method will go away entirely and asset creation will always match the configuration of the snapshot?

Reply:
That's correct - this method will go away with DC-1120.